server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java

```diff
@@ -73,26 +77,62 @@ protected RowRecord nextWithoutConstraint() throws IOException {
    * @return if there has next row record.
    */
   private boolean cacheRowRecord() throws IOException {
-    while (timeGenerator.hasNext()) {
+    Object[][] results = new Object[seriesReaderByTimestampList.size()][];
+    while (timeGenerator.hasNextTimeColumn()) {
+      initResults(results);
+
+      TimeColumn timeColumn = timeGenerator.nextTimeColumn();
+
+      fillResults(results, timeColumn);
+      flatAndCache(results, timeColumn);
+    }
+    hasCachedRowRecord = !cachedRecords.isEmpty();
+    return hasCachedRowRecord;
+  }
+
+  private void flatAndCache(Object[][] results, TimeColumn timeColumn) {
+    int resultSize = timeColumn.size() - timeColumn.position();
+    for (int i = 0; i < resultSize; i++) {
+      RowRecord rowRecord = new RowRecord(timeColumn.getTimeByIndex(i + timeColumn.position()));
```

Comment:
record the position before, avoid calculate each time

Reply:
flatAndCache was combined